### ISSUE TITLE
RFC: global context in async

### ISSUE THREAD

**Issue by janbiedermann** [opal/opal#1956]
Current Situation:
Whenever interfacing asynchronous javascript, opal code ends up with a lot of promises and a chain of .then.then.then. Resulting code is very unlike ruby.
For example (pseude code):
```ruby
net_http.get('http://opal.com').then do |response|
  process(response)
end.then do |processed_response|
  # do something else
end
```

The Idea:
Putting global context in async, thus running opal code within a promise (async), would allow us to write above code more rubyish:
```ruby
response = net_http.get('http://opal.com')
processed_response = process(response)
```

The compiler under the hoods would generate the according javascript instructions, for above code for example something like:
```javascript
var response = await net_http['$get']('http://opal.get');
processed_response = await this['process'](response);
```

The development experience would appear very synchronous, rubyish, although under the hood it is not at all.

And implementing a `sleep 1` would be possible, although it may not behave 100% exactly as the ruby version, it would be a good approximation.

Performance considerations:
Javascript will do all the promise wrapping natively, so it's a fast, but of course not as fast as a simple return of a value. When making every method async by default, this may slow down things.

On the other hand, as javascript will now be able to schedule async things, or even ruby code, more efficiently, as there would be less code which would be required to run in strict sequence, this may result in a overall performance improvement, especially when handling asynchronous calls.

---

So far i have no ideas how this would work with constants, etc. , just would like to share my idea. I think it would be nice if coding in opal could be even more rubyish and less promiseish and that it could clean up some code and would allow for clearer ruby-way thinking.

**Comment by elia** [comment]
Would be great to find a way for this kind of stuff to work, I think it's been tried a bit in https://github.com/opal/opal/issues/1541 without success, not to say it's impossible, but it's good to know what approaches had failed in the past.

**Comment by ahmadine** [comment]
I think the DSL should mimic the Ruby async library

https://github.com/socketry/async

How about doing it this way:
```ruby
Async do
  response = net_http.get('http://opal.com').wait
  processed_response = process(response).wait
end
```

I'm not entirely sold on `Async` being uppercase, but it's how this async library does. And this library doesn't use `wait` everywhere, but I see it more as a feature. I even think that this approach shouldn't need anything more than a simple monkey-patch (may be dependent on some work I started with #2027 to convert between Ruby and JS promises)

**Comment by ahmadine** [comment]
About the monkey-patch, that's not true. I wasn't really able to do it, we need something like #1711, otherwise we are left with very ugly hacks.

**Comment by elia** [comment]
Yes, it needs compiler support. TBH I like the JS.async and JS.await solution, given it doesn't work like normal Ruby. Alternatively, since the underlying implementation of async/await is Promises it might make sense to link the syntax to them… like `promise.await` would be cool, but I'm not sure it can work.

**Comment by ahmadine** [comment]
It can work, but would require a very careful design. Either a new way to define blocks and defs, or something, that causes entire groups of them to be declared async, like the pull request I referenced above. While we are talking about possibilities, I thought maybe about a magic comment that would cause every function in a file to be declared using "async function" and then maybe also this file would be loaded in a special, Promise-concerning way. That kind of seems like an overkill to me (some functions would be declared async and then called in a sync mode...), but... `require "async"` which would define `promise#wait` or `#await`.

I don't know, I would rather want a way to write the code I described above, but that seems to involve adding `Async` as a special language construct.

Oh by the way, does anyone here remember Ruby `Fiber`s? I do, and I learned to hate the API, but the async Gem uses that API, because it is the only way of implementing cooperative multithreading in Ruby. This API is ugly and the implementation is buggy, but I happen to understand it somewhat and if we want to work towards maximum compatibility, we should think about defining it and maybe also implementing the rest of the API in terms of Fibers. But maybe the Fibers are more about the generator functions: the `function* a` syntax. I know that some JS environment polyfill the async functions with generator functions.

When I read: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function* it looks very Fibery to me.

In short, yeah, that's a lot of things to think about and a workable `async/await` implementation is often the thing people consider when choosing their next programming environment.

**Comment by ahmadine** [comment]
I came up with another idea, but I'm not able to implement it yet and implementing it may be a bit tricky.

Let's suppose, that there's a function that returns a native JS Promise (ideally a bridged one). In this case, we could call it this way: `Ajax.get("http://example.com/").JS.await` which would compile to something like: `(await $$(null, "Ajax").$get("http://example.com/"))`.

Now, the second part of the implementation would be to mark such a function/block as "async". This means that it will be defined as `async function funcall()`, obviously, and if someone called that function, it would dispatch not to `funcall()`, but to `(await funcall())`.

The third part would be to define every function that calls any such async function to be marked as async as well and therefore we keep the assumptions of the second part recursively.

The fourth part, this would go down to the lowest level. The main level code, if it (directly) calls `async` functions, would be wrapped in `(async function() {` and `})());`. This is the easiest solution, but we still need to keep order. Therefore we could wrap non-async code in a Promise, async code in `async function`, and then create a promise chain to execute the main level in order.

To think about metaprogramming with that - yeah. This would explicitly mean, that every `send` and every `method_missing` function would become `async` in the process. That's a real downside, unless we make an assumption, that `async` functions can't be used in the metaprogramming context. Or, perhaps, we make a compiler switch, like the one for `method_missing`, but I'm not sold on that. Unless, perhaps, maybe we define something like `wsend` which would check the type of the function first and call it correctly - and `wsend` itself would be an async function. This method could be included in some new stdlib library like `async.rb`.

Inevitably, that will break the ES5 compatibility if any such async function happens to happen in the codebase, but otherwise no (unless we want to keep the `method_missing` assumption).

Furthermore, maybe it would be possible to declare in this potential `async.rb` library an async method `Promise#wait` which would just call `JS.await` on such a promise therefore rubifying the codebase even further.

Why do I think it's a good solution: declaring a function as `async` involves quite a serious strain on the compiler and on the execution. This way we keep `async` functions to the minimum, while enabling the use of them for those that need them. And this solution would keep everything seamless - most people wouldn't even have to bother about the function types, they could keep writing their synchronous Ruby code with no need to think about the implications.

What are the challenges: the Ruby functions are not necessarily required to be declared in a correct order, therefore an additional post-code-generation step with some awareness of the environment (like corelib) would be needed.

**Comment by ahmadine** [comment]
This implementation could be robust enough to later think about implementing `Fiber` and friends, maybe even some https://github.com/socketry/async port/compatibility in terms of that implementation.

**Comment by ahmadine** [comment]
On a second thought, we won't be able to tell if a method is of an async type or not, due to the dynamic typing. For the most time we won't be able to call an async method directly using just `(await funcall())`, it would need to be wrapped in a `wsend` function. Or we could make async all the functions of a given name, but that may be a worse idea.

**Comment by martin-schoettler** [comment]
@hmdne and @elia

I recommend to use the same approach as CoffeeScript 2:

- KISS
- Just create a keyword "await" in Opal, compiling this one-to-one to "await" in JavaScript.
- The Opal compiler should declare all generated JS functions as "async" if (_and only if)_ the "await" keyword is used in the body of the function.

**Reasoning:**

- A solution is _needed_, not only for get() and post() but also for using the Crypto API. (Using then() and catch() for promises is usually a pain.)
- The solution described above is simple and allows to do all which is needed: E. g. net_http.get() can be implemented as a call to an  "async function" if no block is given. Tested sample function:

        async function(
          get_or_post, url, params
        ) {
          // get_or_post is assumed to be 'get' or 'post'
          var fn, promise;
          promise = new Promise(
            ( resolve, reject )=> {
              $[ get_or_post ](
                url,
                params,
                async function( result, status ) {
                  if( status === 'success' ) {
                    resolve( result );
                  } else {
                    reject( status );
                  }
                }
              );
            }
          )
          return await( promise );
        }

- Please keep in mind that "await" operates on _values_, not on functions (returning either the value (if the value is not a promise) or waiting for the non-promise value and returning that). The sample above may be simplified to (tested too):

        function(
          get_or_post, url, params
        ) {
          var fn, promise;
          promise = new Promise(
            ( resolve, reject )=> {
              $[ get_or_post ](
                url,
                params,
                function( result, status ) {
                  if( status === 'success' ) {
                    resolve( result );
                  } else {
                    reject( status );
                  }
                }
              );
            }
          )
          return promise;
        }

  (If the calling code "await"s on the result of a call to that syncronous function, it works as well as an "await" to the async version of the function.)
_Therefore_ it is _not_ required that an awaited function is async. "await function(){ return 1; }" will simply return 1. "await function(){ ...; return some_promise; }" will wait for the promise and then return the resolved value (or throw an exception).

- There are cases, when an async function (or sync function) is called but the result should not be awaited (e. g. HTTP get() just for a logging purpose). Also this case  is not a problem, as the solution described gives the freedom to decide whether to await or not.
- The only case not explicitly covered by the solution described is: The programmer wants to declare a method, which _should_ by asynchronous, but (s)he does not use "await" in the method body. The simple work around would be: Insert a line "await nil" on top of the method body. That should cause the Opal compiler to detect "await" and to make the function async as desired. So no explicite "async" keyword in Opal would be necessary.
- I'm rather sure that "Promise.await( ... )" will not work. Think about "pp = Promise; pp.await( ... )". So I think a _keyword_ "await" is a _must_.
